zerver/tests/test_channel_permissions.py

```diff
@@ -737,6 +737,133 @@ def check_unsubscribing_user(
+        do_change_stream_group_based_setting(
+            stream,
+            "can_unsubscribe_group",
+            anonymous_group_member_dict,
+            acting_user=aaron,
+        )
+        check_unsubscribing_user(aaron)
+
+        owners_group = NamedUserGroup.objects.get(
+            name=SystemGroups.OWNERS, realm=realm, is_system_group=True
+        )
+        self.subscribe(desdemona, stream.name)
+        self.subscribe(aaron, stream.name)
+        do_change_stream_group_based_setting(
+            stream,
+            "can_unsubscribe_group",
+            owners_group,
+            acting_user=admin,
+        )
+        check_unsubscribing_user(desdemona)
+        check_unsubscribing_user(aaron, expect_fail=True)
+
+        hamletcharacters_group = NamedUserGroup.objects.get(name="hamletcharacters", realm=realm)
+        self.subscribe(hamlet, stream.name)
+        do_change_stream_group_based_setting(
+            stream,
+            "can_unsubscribe_group",
+            hamletcharacters_group,
+            acting_user=admin,
+        )
+        check_unsubscribing_user(hamlet)
+        check_unsubscribing_user(aaron, expect_fail=True)
+
+        # Test that an admin can always unsubscribe themselves.
+        self.subscribe(admin, stream.name)
+        check_unsubscribing_user(admin)
+
+        nobody_group = NamedUserGroup.objects.get(
+            name=SystemGroups.NOBODY, realm=realm, is_system_group=True
+        )
+        do_change_stream_group_based_setting(
+            stream,
+            "can_unsubscribe_group",
+            nobody_group,
+            acting_user=admin,
+        )
+
+        # Test that a user in the can_administer_channel_group can
+        # unsubscribe themselves
+        do_change_stream_group_based_setting(
+            stream,
+            "can_administer_channel_group",
+            anonymous_group_member_dict,
+            acting_user=aaron,
+        )
+        check_unsubscribing_user(aaron)
+
+        do_change_stream_group_based_setting(
+            stream,
+            "can_administer_channel_group",
+            nobody_group,
+            acting_user=aaron,
+        )
+
+        # Test that a user in the can_remove_subscribers_group can
+        # unsubscribe themselves
+        self.subscribe(aaron, stream.name)
+        do_change_stream_group_based_setting(
+            stream,
+            "can_remove_subscribers_group",
+            anonymous_group_member_dict,
+            acting_user=aaron,
+        )
+        check_unsubscribing_user(aaron)
+
+        # Test that a user not in any of the permission groups cannot
+        # unsubscribe themselves
+        self.subscribe(cordelia, stream.name)
+        check_unsubscribing_user(cordelia, expect_fail=True)
+
+        # Test that a user is not subscribed to the inaccessible stream.
+        stream = self.make_stream("private_stream", invite_only=True)
+        stream_name_list = [stream.name]
```

Comment:
Would be good to add user to the `can_unsubscribe_group` group as well, even when the user is not subscribed.